File: server/messages.go
package server

type MessageType string

const (
	// both incoming and outcoming
	MessageText      = MessageType("text")
	MessageSetFigure = MessageType("figure")
	// Incoming messages
	MessageStartGame       = MessageType("start")
	MessageCancelSearching = MessageType("cancel")
	MessageCloseGame       = MessageType("close")
	// Outcoming messages
	MessageGameStarted  = MessageType("started")
	MessageYouAreWinner = MessageType("winner")
	MessageYouAreLooser = MessageType("looser")
	MessageDraw         = MessageType("draw")
	MessageGameClosed   = MessageType("closed")
	// messages dedicated to time
	MessageTick    = MessageType("tick")
	MessageTimeout = MessageType("timeout")
)

func SendTextMessageTo(player *Player, message string) {
	player.SendToClient(map[string]any{
		"messageType": MessageText,
		"content":     message,
	})
}

func SendGameStartedMessageTo(player *Player, isPlayingCrosses bool) {
	player.SendToClient(map[string]any{
		"messageType":      MessageGameStarted,
		"isPlayingCrosses": isPlayingCrosses,
	})
}

func SendFigureSetMessageTo(player *Player, row, col int) {
	player.SendToClient(map[string]any{
		"messageType": MessageSetFigure,
		"row":         row,
		"column":      col,
	})
}

func SendWinnerMessageTo(player *Player) {
	player.SendToClient(map[string]any{
		"messageType": MessageYouAreWinner,
	})
}

func SendLooserMessageTo(player *Player) {
	player.SendToClient(map[string]any{
		"messageType": MessageYouAreLooser,
	})
}

func SendDrawMessageTo(player *Player) {
	player.SendToClient(map[string]any{
		"messageType": MessageDraw,
	})
}

func SendOpponentClosedGameTo(player *Player) {
	player.SendToClient(map[string]any{
		"messageType": MessageGameClosed,
	})
}

func SendTickMessageTo(player *Player) {
	player.SendToClient(map[string]any{
		"messageType": MessageTick,
	})
}

func SendTimeoutMessageTo(player *Player) {
	player.SendToClient(map[string]any{
		"messageType": MessageTimeout,
	})
}

File: client/messages.ts
export const enum MessageType {
  // both incoming and outcoming
  TEXT_MESSAGE = 'text',
  SET_FIGURE = 'figure',

  // outcoming
  START_GAME = 'start',
  CANCEL_GAME = 'cancel',
  CLOSE_GAME = 'close',
  
  // incoming
  GAME_STARTED = 'started',
  YOU_ARE_WINNER = 'winner',
  DRAW = 'draw',
  YOU_ARE_LOOSER = 'looser',
  GAME_CLOSED = 'closed',

  // dedicated to time
  TICK = 'tick',
  TIMEOUT = 'timeout'
}

export type OutcomingMessage = 
  | MessageText
  | MessageSetFigure
  | MessageStartGame
  | MessageCancelSearching
  | MessageCloseGame

export type OutcomingMessageType = 
  | OutcomingMessage['messageType']

export type IncomingMessage =
  | MessageText
  | MessageSetFigure
  | MessageGameStarted
  | MessageYouAreWinner
  | MessageDraw
  | MessageYouAreLooser
  | MessageGameClosed
  | MessageTick
  | MessageTimeout

export type IncomingMessageType = 
  | IncomingMessage['messageType']

interface MessageText {
  messageType: MessageType.TEXT_MESSAGE
  content: string
}

interface MessageSetFigure {
  messageType: MessageType.SET_FIGURE
  row: number
  column: number
}

interface MessageStartGame {
  messageType: MessageType.START_GAME
}

interface MessageCancelSearching {
  messageType: MessageType.CANCEL_GAME
}

interface MessageCloseGame {
  messageType: MessageType.CLOSE_GAME
}

interface MessageGameStarted {
  messageType: MessageType.GAME_STARTED
  isPlayingCrosses: boolean
}

interface MessageYouAreWinner {
  messageType: MessageType.YOU_ARE_WINNER
}

interface MessageDraw {
  messageType: MessageType.DRAW
}

interface MessageYouAreLooser {
  messageType: MessageType.YOU_ARE_LOOSER
}

interface MessageGameClosed {
  messageType: MessageType.GAME_CLOSED
}

interface MessageTick {
  messageType: MessageType.TICK
}

interface MessageTimeout {
  messageType: MessageType.TIMEOUT
}